src/Statistics.jl

```diff
@@ -0,0 +1,135 @@
+    copyto!(yold,y)
+    v = zero_tangent_vector.(Ref(M), fill(y,length(x)))
+    for i=1:stop_iter
+        copyto!(yold,y)
+        log!.(Ref(M), v, Ref(yold), x)
+        exp!(M, y, yold, sum( values(w).*v ) / 2 )
+        isapprox(M,y,yold; kwargs...) && break
+    end
+    return y
+end
+
+@doc doc"""
+    median(M,x weights=1/n*ones(n); stop_iter=10000, use_rand = false )
+
+computes the Riemannian median of the vector `x`  of `n` points on the
+[`Manifold`](@ref) `M`. This function is nonsmooth (i.e nondifferentiable) and
+uses a cyclic proximal point scheme. Optionally one can provide
+weights $w_i$ for the weighted Riemannian median. The general formula to compute
+the minimizer reads
+````math
+\argmin_{y\in\mathcal M}\sum_{i=1}^n w_i\mathrm{d}_{\mathcal M}(y,x_i),
+````
+where $\mathrm{d}_{\mathcal M}$ denotes the Riemannian [`distance`](@ref).
+
+The cyclic proximal point can be run with a random cyclic order by setting
+`use_rand` to `true`
+
+Optionally you can provide `x0`, the starting point (by default set to the first
+data point). `stop_iter` denotes the maximal number of iterations to perform and
+the `kwargs...` are passed to [`isapprox`](@ref) to stop, when the inimal change
+between two iterates is small. For more stopping criteria
+check the [`Manopt.jl`](https://manoptjl.org) package and use a solver therefrom.
+
+The algorithm is further described in Algorithm 4.3 and 4.4 in 
+> Bačák, M: Computing Medians and Means in Hadamard Spaces.
+> SIAM Journal on Optimization (2014), 24(3), pp. 1542–1566,
+> doi: [10.1137/140953393](https://doi.org/10.1137/140953393),
+> arxiv: [1210.2145](https://arxiv.org/abs/1210.2145)
+"""
+function median(M::Manifold, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x)); x0=x[1], kwargs...)
+    y = similar_result(M, median, x[1])
+    copyto!(y,x0)
+    return median!(M, y, x; kwargs...)
+end
+@doc doc"""
+    median!(M,y,x,w)
+
+computes the [`median`](@ref) in-place in `y` where the initial value of `y` is
+the starting point of the algorithm.
+"""
+function median!(M::Manifold, y, x::AbstractVector, w::AbstractWeights = weights(ones(length(x)) / length(x));
+            stop_iter=100000,
+            use_random = false,
+            kwargs...
+        ) where {T}
+    n = length(x)
+    yold = similar_result(M,median,y)
+    copyto!(yold,y)
+    order = collect(1:n)
+    v = zero_tangent_vector(M,y)
+    for i=1:stop_iter
+        λ = n/(2*i)
+        copyto!(yold,y)
+        use_random && shuffle!(order)
+        for j=1:n
+            t = min( λ * w[order[j]] / distance(M,y,x[order[j]]) , 1 )
+            log!(M, v, y, x[order[j]])
+            y = exp(M, y, t* v)
+        end
+        isapprox(M, y, yold; kwargs...) && break
+    end
+    return y
+end
+median(M::Euclidean, x::AbstractVector, w::AbstractWeights) = median(x,w)
+median(M::Euclidean, x::AbstractVector) = median(x)
+@doc doc"""
+    var(M,x, corrected::Bool=true, mean=mean(M,x))
```

Comment:
I like the elegance of this signature, but it's probably best to use the one from `StatsBase.var`: https://github.com/JuliaStats/StatsBase.jl/blob/master/src/moments.jl#L41 where these are keyword arguments. They also have `varm` for the case where the mean is known.

Reply:
With the optional arguments I am fine, but `varm` is incorporated here already and they (just) have far more cases and implementations that I think are not that necessary here? They really spend a lot of lines on different cases for `var` and `varm` and I don't feel like copying all of them here.

Reply:
I now did the signatures you proposed below, just have to adapt documentation.